### PR TITLE
[MAINTENANCE] partial `cli` + `usage_stats` typing

### PR DESCRIPTION
**Pull request by Kilo59**
Changes proposed in this pull request:
- lint all of the `cli` package with flake8
   - ignore complexity warnings in order to bring under-linting coverage 
- bring most of the cli package under type checking coverage
- fixed `unexpected keyword argument` bug in `cli.store.store_list`  - https://github.com/great-expectations/great_expectations/pull/6335/files#r1018392354

### CODE REVIEW

File: great_expectations/cli/store.py
Comment (Kilo59) [inline great_expectations/cli/store.py:51]: type-checker caught a real bug/misuse here

File: great_expectations/cli/batch_request.py
Comment (Kilo59) [inline great_expectations/cli/batch_request.py]: This never gets used.
Should we remove this line too? Or is there a reason to ensure that this attribute exists?

File: great_expectations/cli/project.py
Comment (Kilo59) [inline great_expectations/cli/project.py]: This and other similar changes were made because the caller of the method had a value that could have been none.
If needed I also made changes to the downstream methods as needed.
Example: `send_usage_message` in usage stats.

File: great_expectations/cli/batch_request.py
Comment (NathanFarmer) [reply great_expectations/cli/batch_request.py]: This entire method seems like it could be removed in favor of a single error check in the `else`.

File: great_expectations/cli/project.py
Comment (NathanFarmer) [reply great_expectations/cli/project.py]: Why not `Optional[str]`?

File: great_expectations/cli/batch_request.py
Comment (cdkini) [reply great_expectations/cli/batch_request.py]: I want to say we keep it as some sort of check to make sure that property works as intended? we can probably get away from the cast though.

Maybe git blame and let's discuss with whoever wrote it?

File: great_expectations/cli/python_subprocess.py
Comment (cdkini) [inline great_expectations/cli/python_subprocess.py:109]: what about `proc.stdout.readline() or ""`?

File: great_expectations/cli/store.py
Comment (NathanFarmer) [reply great_expectations/cli/store.py:51]: 🙇🏻

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py]: `Optional[str]`?

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py]: `Optional[str]`?

File: great_expectations/cli/toolkit.py
Comment (cdkini) [inline great_expectations/cli/toolkit.py]: Is there a linter or some sort of established way to determine if an import is only being brought in for purposes of type checking? I handrolled something but I'm trying to get away from writing up my own implementations of everything 😆

File: great_expectations/cli/toolkit.py
Comment (cdkini) [inline great_expectations/cli/toolkit.py]: Same note - should we `or ""`?

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py:137]: I don't understand why we landed on this possibly being a string. It would raise an error if it was, so maybe we should just proceed assuming it can't be.

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py]: `Optional[str]`?

File: great_expectations/cli/v012/datasource.py
Comment (cdkini) [inline great_expectations/cli/v012/datasource.py:1085]: yikes

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py]: `Optional[str]`?

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py]: `Optional[str]`?

File: pyproject.toml
Comment (cdkini) [inline pyproject.toml]: You're crushing it 👊🏽

**Review by cdkini — APPROVED** [review]

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py]: Why `return None` above and then assign `None` here?

File: great_expectations/cli/toolkit.py
Comment (NathanFarmer) [inline great_expectations/cli/toolkit.py:1082]: Does `sys.exit` cause the string to be returned? If so, why not handle it?

File: great_expectations/cli/python_subprocess.py
Comment (Kilo59) [reply great_expectations/cli/python_subprocess.py:109]: It's `stdout` that's the problem, not the `.readline()` result.

If `stdout` is `None` we will get an attribute error.

File: great_expectations/cli/v012/datasource.py
Comment (NathanFarmer) [inline great_expectations/cli/v012/datasource.py]: Does this do something in-place?

**Review by NathanFarmer — COMMENTED** [review]
A few questions/comments from me!

File: great_expectations/cli/project.py
Comment (Kilo59) [reply great_expectations/cli/project.py]: Because the word `"optional"` can be confusing to developers.

It isn't "optional"... it could be `None`, but it's never "optional".
`Optional` makes sense for keyword arguments that have a default.

https://fastapi.tiangolo.com/python-types/#using-union-or-optional

File: great_expectations/cli/batch_request.py
Comment (Kilo59) [reply great_expectations/cli/batch_request.py]: @alexsherstinsky 
Although I'd prefer to save runtime changes to another PR (unless there's actually a bug or misuse).

File: great_expectations/cli/v012/datasource.py
Comment (Kilo59) [reply great_expectations/cli/v012/datasource.py]: I don't know.
I assumed it was another implicit error check.

>if we can get a batch then there was no issue.

But it's worth flagging to @alexsherstinsky that the `batch` here are never used or returned, is that intended?

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: In both cases the early `return` means I wouldn't have to do a `type: ignore`, but here the return statements would be far apart which I thought hurt readability.

But I could invert the conditions here and put them next to each other, and then I wouldn't have to ignore errors.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py:1082]: No, `sys.exit` raises a `SystemExit` exception.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: Possibly... I think I do want to get rid of the cast, so I'll either do this or update the types of the called methods that are complaining.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py:137]: I was thinking there is an implicit assumption that this is only a `str` if`data_asset_name` is passed in, in which case it doesn't hit the `elif batch_request` conditional (which would have caused an error). 

Ideally we would encode this assumption by declaring `overload` functions that say...
>if data_asset is None, batch_request can be a string
>if data_asset is not None, batch_request must be a dict

But I was having some minor problems getting it to work as intended and I wasn't even sure if my assumptions there were correct.

https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading

File: great_expectations/cli/project.py
Comment (NathanFarmer) [reply great_expectations/cli/project.py]: Ok I missed that we don't have a default

Comment (Kilo59) [reply great_expectations/cli/project.py]: To clarify for anyone else.
`Union[str, None]` and `Optional[str]` ARE the same thing to the type-checker.
This is just a matter of convention/style.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: It wouldn't surprise me if this existed but I don't know one.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: On second look, in this case I think the `cast` is appropriate. This issue is `click.prompt` doesn't know the `default` return type, so it's declared as an `Any`.

**Review by tannerbeam — COMMENTED** [review]
looks like there's a lot going on here. i'll leave it to the pros to evaluate, but i generally support better type hints, so...nice work!

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: I ended up just returning early. Inverting the conditions made for some kinda hard to read `if not exception_occurred and not icrement_version:`  conditional.
But if folks prefer this I can do that instead.

File: great_expectations/cli/batch_request.py
Comment (alexsherstinsky) [inline great_expectations/cli/batch_request.py:46]: @Kilo59 To help me recall, when do we get an `int` for `additional_batch_request_args`?  Thanks!

File: great_expectations/cli/batch_request.py
Comment (alexsherstinsky) [reply great_expectations/cli/batch_request.py]: @Kilo59 This is very weird and looks like you found a bug.  It says that I authored that line, but GitBlame has been inconsistent in my recent reviews (so I am not sure how this bug ended up in there).  Regardless, could you please try deleting that line that references `datasource.execution_engine` entirely to see if it will work without it?  Thanks!

File: great_expectations/cli/store.py
Comment (alexsherstinsky) [inline great_expectations/cli/store.py:51]: ❤️

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py]: @Kilo59 Would `expectation_suite_name: Optional[str]` not be an equivalent shorter version of the same?  Thanks.

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py:97]: @Kilo59 For what it is worth, in my environment, `PyCharm` is not showing any warnings for the way it currently is.

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py]: @Kilo59 Would `data_asset_name: Optional[str]` not be appropriate shorter version?  Thanks.

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py]: @Kilo59 Again, would the use of `Optional` for `usage_event` be wrong?  Thanks.

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py]: @Kilo59 Same here -- would `Optional` be appropriate?  Thanks.

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py]: @Kilo59 This goes against my preference, because it obscures the fact that `load_checkpoint()` returns a value.  Could we at least perhaps use `_ =` (just to show that in some way)?  Thanks.

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py:358]: @Kilo59 How come we are replacing the use of `Optional` with the longer version of what seems to be the same type declaration?  Thanks.

File: great_expectations/cli/v012/suite.py
Comment (alexsherstinsky) [inline great_expectations/cli/v012/suite.py]: @Kilo59 How about `_ = ...` just to not lose the interpretability that there can be a return value?  Thanks.

File: great_expectations/cli/v012/datasource.py
Comment (alexsherstinsky) [inline great_expectations/cli/v012/datasource.py]: @Kilo59 How about `_ = ...` to show that there is a return value, we are just not using it.  Thanks.

File: great_expectations/cli/v012/datasource.py
Comment (alexsherstinsky) [reply great_expectations/cli/v012/datasource.py]: @Kilo59 If this is a `v012` code, I would leave it as is; we are not supposed to be maintaining it.  In fact, it is prime for full wholesale removal.  There have been a series of JIRA items, focused on removing the V2 codebase; perhaps we can either find and resurrect them or create fresh ones, and bring it up at a meeting.  Thanks!

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [inline great_expectations/cli/toolkit.py]: @Kilo59 How come we are deleting the type declaration?  Thanks.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@Kilo59 I reviewed 12 out of 15 files and noted a few questions; once they are answered, I will finish the entire review.  Thanks for doing this!

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: See the linked conversion
https://github.com/great-expectations/great_expectations/pull/6335#discussion_r1019048012

In summary, because the word `optional` can be a source of confusion.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py:358]: Answered above

File: great_expectations/cli/v012/suite.py
Comment (Kilo59) [reply great_expectations/cli/v012/suite.py]: Will do.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: Because it was wrong. But I can add it back with the correct union.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: See above.

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py:97]: `mypy` should raise a warning with any redundant casts.
https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L24

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: Done

File: great_expectations/cli/v012/datasource.py
Comment (Kilo59) [reply great_expectations/cli/v012/datasource.py]: Done

File: great_expectations/cli/batch_request.py
Comment (Kilo59) [reply great_expectations/cli/batch_request.py]: @alexsherstinsky I really don't want to get into changing run time behavior in this PR.
This was meant to be a quick PR to bring some of this CLI code under type coverage.

File: great_expectations/cli/batch_request.py
Comment (Kilo59) [reply great_expectations/cli/batch_request.py:46]: Because of the `limit` keyword to `BatchRequest`
https://github.com/great-expectations/great_expectations/blob/79c6ed9f1fd944bdc96369fcbfdcd969f0f481bc/great_expectations/core/batch.py#L408

File: great_expectations/cli/toolkit.py
Comment (Kilo59) [reply great_expectations/cli/toolkit.py]: Added back.

File: great_expectations/cli/batch_request.py
Comment (alexsherstinsky) [reply great_expectations/cli/batch_request.py:46]: Oh yes.  Thanks.

File: great_expectations/cli/toolkit.py
Comment (alexsherstinsky) [reply great_expectations/cli/toolkit.py]: @Kilo59 I feel that diverse style is even more confusion.  I think that we either use `Optional` everywhere or use the `Union` with `None` everywhere.  To me, consistency is more important than which alternatives we adopt.  Hence, until we have agreed upon and implemented a consistent approach across the entire codebase, I would feel better about keeping `Optional`.  Thanks.

**Comment by alexsherstinsky** [comment]
@Kilo59 I am provisionally approving that you fix the failing tests (obviously) and consider my follow-up comment (otherwise, good stuff!).  Thanks!

**Review by alexsherstinsky — APPROVED** [review]
LGTM (pending fixing tests and considering the follow up suggesting regarding the use of the `Optional` type).  Thank you!

**Pull request merged**
